linkerd/proxy/pool/src/worker.rs

```diff
@@ -0,0 +1,349 @@
+                    biased;
+
+                    // If either the discovery stream or the pool fail, close
+                    // the request stream and process any remaining requests.
+                    e = worker.discover_while_awaiting_requests(), if terminal_failure.is_none() => {
+                        let err = TerminalFailure::new(e);
+                        terminate.take().expect("must not fail twice").send(err.clone());
+                        reqs_rx.close();
+                        tracing::trace!("Closed");
+                        terminal_failure = Some(err);
+                        continue;
+                    }
+
+                    msg = reqs_rx.recv() => match msg {
+                        Some(msg) => msg,
+                        None => {
+                            tracing::debug!("Requests channel closed");
+                            return Ok(());
+                        }
+                    },
+                };
+
+                // Preserve the original request's tracing context.
+                let _enter = span.enter();
+
+                // Wait for the pool to have at least one ready endpoint.
+                if terminal_failure.is_none() {
+                    tracing::trace!("Waiting for pool");
+                    if let Err(e) = worker.ready_pool().await {
+                        let err = TerminalFailure::new(e);
+                        terminate
+                            .take()
+                            .expect("must not fail twice")
+                            .send(err.clone());
+                        reqs_rx.close();
+                        terminal_failure = Some(err);
+                        tracing::trace!("Closed");
+                    } else {
+                        tracing::trace!("Pool ready");
+                    }
+                }
+
+                // Process requests, either by dispatching them to the pool or
+                // by serving errors directly.
+                let _ = if let Some(e) = terminal_failure.clone() {
+                    tx.send(Err(e.into()))
+                } else {
+                    tx.send(worker.pool.call(req))
+                };
+
+                // TODO(ver) track histogram from t0 until the request is dispatched.
+                tracing::debug!(
+                    latency = (time::Instant::now() - t0).as_secs_f64(),
+                    "Dispatched"
+                );
+            }
+        }
+        .instrument(debug_span!("pool")),
+    )
+}
+
+// === impl Worker ===
+
+impl<T, R, P> Worker<R, P>
+where
+    T: Clone + Eq + std::fmt::Debug,
+    R: TryStream<Ok = Update<T>> + Unpin,
+    R::Error: Into<Error>,
+{
+    /// Attempts to update the pool with discovery updates.
+    ///
+    /// Additionally, this attempts to drive the pool to ready if it is
+    /// currently in failfast.
+    ///
+    /// If the discovery stream is closed, this never returns.
```

Comment:
maybe worth documenting that this only returns on errors, and that it's expected to only ever be called in a select with the request channel?